src/Microsoft.AspNetCore.Mvc.Core/ApplicationParts/AssemblyPartDiscoveryModel.cs

```diff
@@ -0,0 +1,188 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+
+namespace Microsoft.AspNetCore.Mvc.ApplicationParts
+{
+    public class AssemblyPartDiscoveryModel : IEquatable<AssemblyPartDiscoveryModel>
+    {
+        internal static readonly string[] ViewsAssemblySuffixes = new[]
+        {
+            ".PrecompiledViews",
+            ".Views",
+        };
+
+        private readonly List<AssemblyPartDiscoveryModel> _additionalParts;
+
+        public AssemblyPartDiscoveryModel(Assembly assembly)
+            : this(assembly, assembly.GetName().Name, assembly.GetCustomAttributes(inherit: false))
+        {
+        }
+
+        public AssemblyPartDiscoveryModel(
+            Assembly assembly, 
+            string name,
+            IReadOnlyList<object> attributes)
+        {
+            Assembly = assembly;
+            Name = name;
+            Attributes = attributes;
+            _additionalParts = new List<AssemblyPartDiscoveryModel>();
+        }
+
+        public string Name { get; }
+
+        public Assembly Assembly { get; }
+
+        public IReadOnlyList<object> Attributes { get; }
+
+        public AssemblyPartDiscoveryModel Parent { get; private set; }
+
+        public IReadOnlyList<AssemblyPartDiscoveryModel> AdditionalParts => _additionalParts;
+
+        public void AddAdditionalPartModel(AssemblyPartDiscoveryModel additionalPart)
+        {
+            _additionalParts.Add(additionalPart);
+            additionalPart.Parent = this;
+        }
+
+        public ApplicationPart ToApplicationPart()
+        {
+            for (var i = 0; i < ViewsAssemblySuffixes.Length; i++)
+            {
+                if (Name.EndsWith(ViewsAssemblySuffixes[i], StringComparison.OrdinalIgnoreCase))
+                {
+                    return new CompiledViewsApplicationPart(Assembly);
+                }
+            }
+
+            return new AssemblyPart(Assembly);
+        }
+
+        public override string ToString() => Name;
+
+        public static AssemblyPartDiscoveryModel ResolveEntryPoint(Assembly entryAssembly)
```

Comment:
How much of this stuff is abstaction, how much is implementation and how much is data?

Looking at the method bodies here there's a lot of **public** surface area that is the implementation. It's doin me a heckin concern

Reply:
I should be able to move all of this API to an internal type. The only user-relevant abstraction here is the actual model type.